File: ruby/hecks/runtime/shell_dispatcher.rb
require "open3"
require "json"

module Hecks
  class Runtime

    # Hecks::Runtime::ShellDispatcher
    #
    # Invokes a Hecksagon::Structure::ShellAdapter with runtime-provided
    # attributes, substitutes {{placeholder}} tokens in the arg vector,
    # executes via Open3.capture3 (never through a shell), parses stdout
    # according to the adapter's output_format, and returns a Result.
    #
    # Security:
    # - command is a fixed binary (Structure validates no "{{" in it)
    # - args are always an array of strings — placeholders substitute
    #   per-element; there is no shell string form
    # - env is empty by default; only entries declared on the adapter
    #   are passed through
    # - working_dir resolves from the adapter's own working_dir (caller
    #   is responsible for making it absolute before constructing the
    #   adapter, or Dir.pwd is used when it's nil)
    # - stdin is unused in v1 (no piping)
    #
    #   result = ShellDispatcher.call(adapter, range: "HEAD~5..HEAD")
    #   result.output       # => format-parsed
    #   result.raw_stdout   # => raw string
    #   result.stderr       # => raw stderr string
    #   result.exit_status  # => Integer
    #
    module ShellDispatcher
      # Return shape for a successful (or :exit_code-format) dispatch.
      Result = Struct.new(:output, :raw_stdout, :stderr, :exit_status, keyword_init: true)

      module_function

      # Dispatch a shell adapter.
      #
      # @param adapter [Hecksagon::Structure::ShellAdapter]
      # @param attrs [Hash{Symbol=>Object}] values substituted into {{placeholders}}
      # @return [Result]
      # @raise [Hecks::ShellAdapterError] if exit status is non-zero (except :exit_code format)
      # @raise [Hecks::ShellAdapterTimeoutError] if adapter.timeout is exceeded
      def call(adapter, attrs = {})
        substituted = substitute_placeholders(adapter.args, attrs)
        run(adapter, substituted)
      end

      # Substitute {{name}} tokens in each arg element from +attrs+.
      # Unknown placeholders are left as-is so callers see the literal
      # token in the executed command (helpful for debugging).
      #
      # @param args [Array<String>]
      # @param attrs [Hash]
      # @return [Array<String>]
      def substitute_placeholders(args, attrs)
        sym_attrs = attrs.transform_keys(&:to_sym)
        args.map do |arg|
          arg.gsub(Hecksagon::Structure::ShellAdapter::PLACEHOLDER_RE) do
            name = Regexp.last_match(1).to_sym
            sym_attrs.key?(name) ? sym_attrs[name].to_s : Regexp.last_match(0)
          end
        end
      end

      # Execute the adapter and return a Result.
      #
      # @param adapter [Hecksagon::Structure::ShellAdapter]
      # @param args [Array<String>]
      # @return [Result]
      def run(adapter, args)
        chdir = adapter.working_dir || Dir.pwd
        env = adapter.env
        stdout, stderr, status = capture(adapter, env, args, chdir)
        exit_status = status.respond_to?(:exitstatus) ? status.exitstatus : status.to_i

        if exit_status != 0 && adapter.output_format != :exit_code
          short = stderr.to_s.strip.lines.first.to_s[0, 80]
          short += "…" if stderr.to_s.strip.length > short.length
          raise Hecks::ShellAdapterError.new(
            "shell adapter :#{adapter.name} exited #{exit_status}: #{short}",
            adapter: adapter.name, exit_status: exit_status, stderr: stderr
          )
        end

        Result.new(
          output: parse_output(adapter.output_format, stdout, exit_status),
          raw_stdout: stdout,
          stderr: stderr,
          exit_status: exit_status
        )
      end

      # Capture stdout/stderr/status with optional timeout wrapping.
      #
      # @return [Array(String, String, Process::Status)]
      def capture(adapter, env, args, chdir)
        # Open3.capture3 with unsetenv_others clears the parent env first,
        # then layers adapter.env on top. Stdin is always closed empty.
        opts = {
          chdir: chdir,
          stdin_data: "",
          unsetenv_others: true
        }
        return Open3.capture3(env, adapter.command, *args, **opts) unless adapter.timeout
        capture_with_timeout(adapter, env, args, chdir)
      end

      # Spawn with a timeout, actively killing the child (and its process
      # group) so the caller doesn't wait the full child duration just
      # because the dispatcher raised early.
      #
      # @return [Array(String, String, Process::Status)]
      def capture_with_timeout(adapter, env, args, chdir)
        stdin, stdout, stderr, wait_thr = Open3.popen3(
          env, adapter.command, *args,
          chdir: chdir,
          unsetenv_others: true,
          pgroup: true
        )
        stdin.close
        pid = wait_thr.pid
        deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + adapter.timeout.to_f

        until wait_thr.join(0.01)
          if Process.clock_gettime(Process::CLOCK_MONOTONIC) >= deadline
            begin
              Process.kill("-KILL", pid)
            rescue Errno::ESRCH, Errno::EPERM
              # Process already exited or pgroup kill refused
            end
            wait_thr.join
            stdout.close rescue nil
            stderr.close rescue nil
            # streams intentionally not read — killed process output is discarded
            raise Hecks::ShellAdapterTimeoutError.new(
              "shell adapter :#{adapter.name} exceeded #{adapter.timeout}s timeout",
              adapter: adapter.name, timeout: adapter.timeout
            )
          end
        end
        [stdout.read, stderr.read, wait_thr.value]
      ensure
        [stdin, stdout, stderr].each { |io| io.close if io && !io.closed? }
      end

      # Parse stdout by output_format. Exit-code format discards stdout
      # and returns the exit status as an Integer.
      #
      # @return [Object] format-specific output
      def parse_output(format, stdout, exit_status)
        case format
        when :text       then stdout
        when :lines      then stdout.to_s.each_line.map(&:chomp).reject(&:empty?)
        when :json       then JSON.parse(stdout.to_s)
        when :json_lines then stdout.to_s.each_line.reject { |l| l.strip.empty? }.map { |l| JSON.parse(l) }
        when :exit_code  then exit_status
        else raise ArgumentError, "unknown output_format: #{format.inspect}"
        end
      end
    end
  end
end

File: rust/src/runtime/shell_dispatcher.rs
//! ShellDispatcher — Rust mirror of lib/hecks/runtime/shell_dispatcher.rb
//!
//! Takes a ShellAdapter from the parsed hecksagon and attrs from the
//! caller, substitutes `{{placeholder}}` tokens into the arg vector,
//! then executes via std::process::Command (no shell).
//!
//! Security (parity with Ruby):
//!   * command is a fixed binary (parser validates no `{{` in it)
//!   * args are per-element — placeholder substitution happens per arg,
//!     no shell-string form, no word splitting
//!   * env starts empty — only entries declared on the adapter pass
//!     through (env_clear equivalent of Ruby's `unsetenv_others: true`)
//!   * stdin is empty
//!   * timeout uses thread-based polling with SIGKILL on expiry
//!     (mirrors Ruby's Process.kill("-KILL", pid))
//!
//! Output formats (parity):
//!   :text       → raw stdout String
//!   :lines      → Vec<String>, chomped, empty lines dropped
//!   :json       → serde_json::Value parsed from full stdout
//!   :json_lines → Vec<Value> one per non-empty line
//!   :exit_code  → i32 (stdout discarded)

use crate::hecksagon_ir::ShellAdapter;
use crate::runtime::adapter_io::substitute;
use std::collections::HashMap;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

/// Return shape for a successful dispatch (parity with Ruby's Result
/// struct in shell_dispatcher.rb).
#[derive(Debug, Clone)]
pub struct Result {
    pub output: Output,
    pub raw_stdout: String,
    pub stderr: String,
    pub exit_status: i32,
}

#[derive(Debug, Clone)]
pub enum Output {
    Text(String),
    Lines(Vec<String>),
    Json(serde_json::Value),
    JsonLines(Vec<serde_json::Value>),
    ExitCode(i32),
}

#[derive(Debug)]
pub enum DispatchError {
    /// Non-zero exit when output_format != :exit_code.
    NonZeroExit { adapter: String, exit_status: i32, stderr: String },
    /// Process still alive after `adapter.timeout` seconds; sent SIGKILL.
    Timeout { adapter: String, seconds: u64 },
    /// Spawn failed (binary not found, permission, etc).
    SpawnFailed { adapter: String, source: String },
    /// output_format parser could not read the bytes.
    ParseFailed { adapter: String, format: String, message: String },
}

impl std::fmt::Display for DispatchError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DispatchError::NonZeroExit { adapter, exit_status, stderr } =>
                write!(f, "shell adapter :{} exited {}: {}", adapter, exit_status, stderr.trim()),
            DispatchError::Timeout { adapter, seconds } =>
                write!(f, "shell adapter :{} exceeded {}s timeout", adapter, seconds),
            DispatchError::SpawnFailed { adapter, source } =>
                write!(f, "shell adapter :{} spawn failed: {}", adapter, source),
            DispatchError::ParseFailed { adapter, format, message } =>
                write!(f, "shell adapter :{} output_format={} parse failed: {}", adapter, format, message),
        }
    }
}

/// Dispatch a shell adapter with the given attribute bindings.
pub fn call(adapter: &ShellAdapter, attrs: &HashMap<String, String>) -> std::result::Result<Result, DispatchError> {
    let substituted: Vec<String> = adapter.args.iter().map(|a| substitute(a, attrs)).collect();
    run(adapter, &substituted)
}

fn run(adapter: &ShellAdapter, args: &[String]) -> std::result::Result<Result, DispatchError> {
    let working_dir = adapter.working_dir.clone()
        .unwrap_or_else(|| std::env::current_dir().map(|p| p.to_string_lossy().into()).unwrap_or_else(|_| ".".into()));
    let mut cmd = Command::new(&adapter.command);
    cmd.args(args)
       .current_dir(&working_dir)
       .env_clear()
       .stdin(Stdio::null())
       .stdout(Stdio::piped())
       .stderr(Stdio::piped());
    for (k, v) in &adapter.env {
        cmd.env(k, v);
    }
    let (stdout_bytes, stderr_bytes, exit_status) = match adapter.timeout {
        None => spawn_and_wait(adapter, cmd)?,
        Some(secs) => spawn_with_timeout(adapter, cmd, Duration::from_secs(secs))?,
    };
    let stdout = String::from_utf8_lossy(&stdout_bytes).into_owned();
    let stderr = String::from_utf8_lossy(&stderr_bytes).into_owned();

    if exit_status != 0 && adapter.output_format != "exit_code" {
        return Err(DispatchError::NonZeroExit {
            adapter: adapter.name.clone(),
            exit_status,
            stderr,
        });
    }
    let output = parse_output(&adapter.output_format, &stdout, exit_status)
        .map_err(|e| DispatchError::ParseFailed {
            adapter: adapter.name.clone(),
            format: adapter.output_format.clone(),
            message: e,
        })?;
    Ok(Result { output, raw_stdout: stdout, stderr, exit_status })
}

fn spawn_and_wait(adapter: &ShellAdapter, mut cmd: Command)
    -> std::result::Result<(Vec<u8>, Vec<u8>, i32), DispatchError>
{
    let out = cmd.output().map_err(|e| DispatchError::SpawnFailed {
        adapter: adapter.name.clone(),
        source: e.to_string(),
    })?;
    Ok((out.stdout, out.stderr, out.status.code().unwrap_or(-1)))
}

/// Timeout path — poll the child; SIGKILL on expiry. Matches the Ruby
/// `capture_with_timeout` loop that actively kills pgroup on deadline.
fn spawn_with_timeout(adapter: &ShellAdapter, mut cmd: Command, timeout: Duration)
    -> std::result::Result<(Vec<u8>, Vec<u8>, i32), DispatchError>
{
    use std::io::Read;
    let mut child = cmd.spawn().map_err(|e| DispatchError::SpawnFailed {
        adapter: adapter.name.clone(),
        source: e.to_string(),
    })?;
    let mut stdout = child.stdout.take().expect("piped stdout");
    let mut stderr = child.stderr.take().expect("piped stderr");
    let deadline = Instant::now() + timeout;
    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                let mut out = Vec::new(); let mut err = Vec::new();
                let _ = stdout.read_to_end(&mut out);
                let _ = stderr.read_to_end(&mut err);
                return Ok((out, err, status.code().unwrap_or(-1)));
            }
            Ok(None) => {
                if Instant::now() >= deadline {
                    let _ = child.kill();
                    let _ = child.wait();
                    return Err(DispatchError::Timeout {
                        adapter: adapter.name.clone(),
                        seconds: timeout.as_secs(),
                    });
                }
                std::thread::sleep(Duration::from_millis(10));
            }
            Err(e) => return Err(DispatchError::SpawnFailed {
                adapter: adapter.name.clone(),
                source: e.to_string(),
            }),
        }
    }
}

fn parse_output(format: &str, stdout: &str, exit_status: i32)
    -> std::result::Result<Output, String>
{
    match format {
        "text" => Ok(Output::Text(stdout.to_string())),
        "lines" => Ok(Output::Lines(
            stdout.lines().filter(|l| !l.is_empty()).map(|l| l.to_string()).collect()
        )),
        "json" => serde_json::from_str(stdout)
            .map(Output::Json)
            .map_err(|e| e.to_string()),
        "json_lines" => {
            let mut out = Vec::new();
            for line in stdout.lines().filter(|l| !l.trim().is_empty()) {
                let v: serde_json::Value = serde_json::from_str(line).map_err(|e| e.to_string())?;
                out.push(v);
            }
            Ok(Output::JsonLines(out))
        }
        "exit_code" => Ok(Output::ExitCode(exit_status)),
        other => Err(format!("unknown output_format: {}", other)),
    }
}
